File: examples/Algorithm/python/14_shortest_path.py
"""
Dijkstra's Algorithm
Dijkstra's Shortest Path Algorithm

Finds the single-source shortest path in a weighted graph.
Used for graphs without negative weights.
"""

import heapq
from collections import defaultdict
from typing import List, Dict, Tuple, Optional


# =============================================================================
# Weighted Graph Representation
# =============================================================================
def create_weighted_graph(edges: List[Tuple[int, int, int]], directed: bool = False) -> Dict[int, List[Tuple[int, int]]]:
    """
    Create weighted adjacency list from edge list (u, v, weight)
    graph[u] = [(v, weight), ...]
    """
    graph = defaultdict(list)
    for u, v, w in edges:
        graph[u].append((v, w))
        if not directed:
            graph[v].append((u, w))
    return graph


# =============================================================================
# 1. Basic Dijkstra Implementation
# =============================================================================
def dijkstra(graph: Dict[int, List[Tuple[int, int]]], start: int, n: int) -> List[int]:
    """
    Dijkstra's Algorithm (using priority queue)
    Time Complexity: O((V + E) log V)

    Args:
        graph: Adjacency list (node -> [(neighbor, weight), ...])
        start: Starting node
        n: Total number of nodes

    Returns:
        Array of shortest distances to each node (infinity if unreachable)
    """
    INF = float('inf')
    dist = [INF] * n
    dist[start] = 0

    # Min heap storing (distance, node) tuples
    pq = [(0, start)]

    while pq:
        d, u = heapq.heappop(pq)

        # Skip if already processed with shorter distance
        if d > dist[u]:
            continue

        # Check adjacent nodes
        for v, weight in graph[u]:
            new_dist = dist[u] + weight
            if new_dist < dist[v]:
                dist[v] = new_dist
                heapq.heappush(pq, (new_dist, v))

    return dist


# =============================================================================
# 2. Dijkstra + Path Tracking
# =============================================================================
def dijkstra_with_path(
    graph: Dict[int, List[Tuple[int, int]]],
    start: int,
    end: int,
    n: int
) -> Tuple[int, List[int]]:
    """
    Returns shortest distance along with the actual path
    """
    INF = float('inf')
    dist = [INF] * n
    dist[start] = 0
    parent = [-1] * n

    pq = [(0, start)]

    while pq:
        d, u = heapq.heappop(pq)

        if d > dist[u]:
            continue

        if u == end:
            break

        for v, weight in graph[u]:
            new_dist = dist[u] + weight
            if new_dist < dist[v]:
                dist[v] = new_dist
                parent[v] = u
                heapq.heappush(pq, (new_dist, v))

    # Reconstruct path
    if dist[end] == INF:
        return INF, []

    path = []
    node = end
    while node != -1:
        path.append(node)
        node = parent[node]
    path.reverse()

    return dist[end], path


# =============================================================================
# 3. All-Pairs Shortest Path (Floyd-Warshall)
# =============================================================================
def floyd_warshall(n: int, edges: List[Tuple[int, int, int]]) -> List[List[int]]:
    """
    Floyd-Warshall Algorithm
    Computes shortest distances between all pairs of vertices
    Time Complexity: O(V^3)
    """
    INF = float('inf')
    dist = [[INF] * n for _ in range(n)]

    # Distance to self is 0
    for i in range(n):
        dist[i][i] = 0

    # Apply edge information
    for u, v, w in edges:
        dist[u][v] = w

    # Consider paths through vertex k
    for k in range(n):
        for i in range(n):
            for j in range(n):
                if dist[i][k] + dist[k][j] < dist[i][j]:
                    dist[i][j] = dist[i][k] + dist[k][j]

    return dist


# =============================================================================
# 4. Bellman-Ford Algorithm (Allows Negative Weights)
# =============================================================================
def bellman_ford(n: int, edges: List[Tuple[int, int, int]], start: int) -> Tuple[List[int], bool]:
    """
    Bellman-Ford Algorithm
    Allows negative weights and detects negative cycles
    Time Complexity: O(VE)

    Returns:
        (shortest distance array, whether negative cycle exists)
    """
    INF = float('inf')
    dist = [INF] * n
    dist[start] = 0

    # Repeat V-1 times
    for _ in range(n - 1):
        for u, v, w in edges:
            if dist[u] != INF and dist[u] + w < dist[v]:
                dist[v] = dist[u] + w

    # Detect negative cycle (if update occurs on one more iteration, negative cycle exists)
    has_negative_cycle = False
    for u, v, w in edges:
        if dist[u] != INF and dist[u] + w < dist[v]:
            has_negative_cycle = True
            break

    return dist, has_negative_cycle


# =============================================================================
# 5. Network Delay Time Problem
# =============================================================================
def network_delay_time(times: List[List[int]], n: int, k: int) -> int:
    """
    Minimum time for signal to reach all nodes from node k
    times[i] = [source, target, time]
    Returns -1 if not all nodes are reachable
    """
    graph = defaultdict(list)
    for u, v, w in times:
        graph[u].append((v, w))

    dist = dijkstra(graph, k, n + 1)  # Node numbers start from 1

    # Maximum distance among nodes 1~n
    max_time = max(dist[1:n + 1])

    return max_time if max_time != float('inf') else -1


# =============================================================================
# 6. Kth Shortest Path
# =============================================================================
def kth_shortest_path(
    graph: Dict[int, List[Tuple[int, int]]],
    start: int,
    end: int,
    k: int,
    n: int
) -> int:
    """
    Return the length of the kth shortest path
    Returns -1 if not found
    """
    INF = float('inf')
    count = [0] * n  # Number of arrivals at each node
    pq = [(0, start)]

    while pq:
        d, u = heapq.heappop(pq)
        count[u] += 1

        if u == end and count[u] == k:
            return d

        # Don't expand nodes visited more than k times
        if count[u] > k:
            continue

        for v, weight in graph[u]:
            heapq.heappush(pq, (d + weight, v))

    return -1


# =============================================================================
# Tests
# =============================================================================
def main():
    print("=" * 60)
    print("Dijkstra & Shortest Path Algorithms")
    print("=" * 60)

    # Example graph
    #       1
    #    0 ---> 1
    #    |      |
    #  4 |      | 2
    #    v      v
    #    2 ---> 3
    #       3
    edges = [
        (0, 1, 1),
        (0, 2, 4),
        (1, 3, 2),
        (2, 3, 3),
        (1, 2, 2)
    ]

    print("\n[Graph Structure]")
    print("    0 --1--> 1")
    print("    |        |")
    print("    4        2")
    print("    v        v")
    print("    2 --3--> 3")
    print("    (1->2 weight 2)")

    # 1. Basic Dijkstra
    print("\n[1] Basic Dijkstra")
    graph = create_weighted_graph(edges, directed=True)
    dist = dijkstra(graph, 0, 4)
    print(f"    Start: 0")
    for i, d in enumerate(dist):
        print(f"    Distance to node {i}: {d}")

    # 2. Dijkstra + Path
    print("\n[2] Dijkstra + Path Tracking")
    distance, path = dijkstra_with_path(graph, 0, 3, 4)
    print(f"    0 -> 3 shortest distance: {distance}")
    print(f"    Path: {' -> '.join(map(str, path))}")

    # 3. Floyd-Warshall
    print("\n[3] Floyd-Warshall (All-Pairs Shortest Distance)")
    all_dist = floyd_warshall(4, edges)
    print("    Distance matrix:")
    for i, row in enumerate(all_dist):
        row_str = [str(d) if d != float('inf') else 'inf' for d in row]
        print(f"    {i}: {row_str}")

    # 4. Bellman-Ford
    print("\n[4] Bellman-Ford (Allows Negative Weights)")
    edges_negative = [(0, 1, 4), (0, 2, 5), (1, 2, -3), (2, 3, 4)]
    dist, has_neg_cycle = bellman_ford(4, edges_negative, 0)
    print(f"    Edges: {edges_negative}")
    print(f"    Shortest distances: {dist}")
    print(f"    Negative cycle: {has_neg_cycle}")

    # Negative cycle example
    edges_neg_cycle = [(0, 1, 1), (1, 2, -1), (2, 0, -1)]
    dist, has_neg_cycle = bellman_ford(3, edges_neg_cycle, 0)
    print(f"\n    Negative cycle graph: {edges_neg_cycle}")
    print(f"    Negative cycle exists: {has_neg_cycle}")

    # 5. Network Delay Time
    print("\n[5] Network Delay Time")
    times = [[2, 1, 1], [2, 3, 1], [3, 4, 1]]
    n, k = 4, 2
    result = network_delay_time(times, n, k)
    print(f"    times={times}, n={n}, k={k}")
    print(f"    Time to reach all nodes: {result}")

    # 6. Kth Shortest Path
    print("\n[6] Kth Shortest Path")
    edges_k = [(0, 1, 1), (0, 2, 3), (1, 2, 1), (1, 3, 2), (2, 3, 1)]
    graph_k = create_weighted_graph(edges_k, directed=True)
    for k in range(1, 4):
        dist = kth_shortest_path(graph_k, 0, 3, k, 4)
        print(f"    0->3 {k}th shortest path: {dist}")

    print("\n" + "=" * 60)
    print("Algorithm Comparison")
    print("=" * 60)
    print("""
    | Algorithm      | Time Complexity | Neg Weights | Use Case              |
    |---------------|----------------|-------------|------------------------|
    | Dijkstra      | O((V+E)log V)  | No          | Single-source shortest |
    | Bellman-Ford  | O(VE)          | Yes         | Neg weights/cycles     |
    | Floyd-Warshall| O(V^3)         | Yes*        | All-pairs shortest     |

    * Floyd-Warshall is also incorrect with negative cycles
    """)


if __name__ == "__main__":
    main()

File: examples/Algorithm/cpp/14_shortest_path.cpp
/*
 * Shortest Path
 * Dijkstra, Bellman-Ford, Floyd-Warshall, 0-1 BFS
 *
 * Algorithms for finding shortest distances between vertices in a graph.
 */

#include <iostream>
#include <vector>
#include <queue>
#include <climits>
#include <algorithm>

using namespace std;

const int INF = INT_MAX;

// =============================================================================
// 1. Dijkstra's Algorithm
// =============================================================================

vector<int> dijkstra(int n, const vector<vector<pair<int, int>>>& adj, int start) {
    vector<int> dist(n, INF);
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<>> pq;

    dist[start] = 0;
    pq.push({0, start});

    while (!pq.empty()) {
        auto [d, u] = pq.top();
        pq.pop();

        if (d > dist[u]) continue;

        for (auto [v, w] : adj[u]) {
            if (dist[u] + w < dist[v]) {
                dist[v] = dist[u] + w;
                pq.push({dist[v], v});
            }
        }
    }

    return dist;
}

// Path reconstruction
pair<vector<int>, vector<int>> dijkstraWithPath(int n, const vector<vector<pair<int, int>>>& adj, int start) {
    vector<int> dist(n, INF);
    vector<int> parent(n, -1);
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<>> pq;

    dist[start] = 0;
    pq.push({0, start});

    while (!pq.empty()) {
        auto [d, u] = pq.top();
        pq.pop();

        if (d > dist[u]) continue;

        for (auto [v, w] : adj[u]) {
            if (dist[u] + w < dist[v]) {
                dist[v] = dist[u] + w;
                parent[v] = u;
                pq.push({dist[v], v});
            }
        }
    }

    return {dist, parent};
}

vector<int> reconstructPath(int start, int end, const vector<int>& parent) {
    vector<int> path;
    for (int at = end; at != -1; at = parent[at]) {
        path.push_back(at);
    }
    reverse(path.begin(), path.end());

    if (path[0] != start) {
        return {};  // No path
    }

    return path;
}

// =============================================================================
// 2. Bellman-Ford Algorithm
// =============================================================================

struct Edge {
    int from, to, weight;
};

pair<vector<int>, bool> bellmanFord(int n, const vector<Edge>& edges, int start) {
    vector<int> dist(n, INF);
    dist[start] = 0;

    // n-1 relaxations
    for (int i = 0; i < n - 1; i++) {
        bool updated = false;
        for (const auto& e : edges) {
            if (dist[e.from] != INF && dist[e.from] + e.weight < dist[e.to]) {
                dist[e.to] = dist[e.from] + e.weight;
                updated = true;
            }
        }
        if (!updated) break;  // Early termination
    }

    // Negative cycle check
    for (const auto& e : edges) {
        if (dist[e.from] != INF && dist[e.from] + e.weight < dist[e.to]) {
            return {dist, true};  // Negative cycle exists
        }
    }

    return {dist, false};
}

// =============================================================================
// 3. Floyd-Warshall Algorithm
// =============================================================================

vector<vector<int>> floydWarshall(int n, vector<vector<int>>& dist) {
    // dist is the initial adjacency matrix (INF if no connection)

    for (int k = 0; k < n; k++) {
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (dist[i][k] != INF && dist[k][j] != INF) {
                    dist[i][j] = min(dist[i][j], dist[i][k] + dist[k][j]);
                }
            }
        }
    }

    return dist;
}

// Version with path reconstruction
pair<vector<vector<int>>, vector<vector<int>>> floydWarshallWithPath(int n, vector<vector<int>>& dist) {
    vector<vector<int>> next(n, vector<int>(n, -1));

    // Initialize: directly connected cases
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            if (i != j && dist[i][j] != INF) {
                next[i][j] = j;
            }
        }
    }

    for (int k = 0; k < n; k++) {
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (dist[i][k] != INF && dist[k][j] != INF &&
                    dist[i][k] + dist[k][j] < dist[i][j]) {
                    dist[i][j] = dist[i][k] + dist[k][j];
                    next[i][j] = next[i][k];
                }
            }
        }
    }

    return {dist, next};
}

// =============================================================================
// 4. 0-1 BFS
// =============================================================================

vector<int> bfs01(int n, const vector<vector<pair<int, int>>>& adj, int start) {
    vector<int> dist(n, INF);
    deque<int> dq;

    dist[start] = 0;
    dq.push_front(start);

    while (!dq.empty()) {
        int u = dq.front();
        dq.pop_front();

        for (auto [v, w] : adj[u]) {
            if (dist[u] + w < dist[v]) {
                dist[v] = dist[u] + w;
                if (w == 0) {
                    dq.push_front(v);
                } else {
                    dq.push_back(v);
                }
            }
        }
    }

    return dist;
}

// =============================================================================
// 5. SPFA (Shortest Path Faster Algorithm)
// =============================================================================

pair<vector<int>, bool> spfa(int n, const vector<vector<pair<int, int>>>& adj, int start) {
    vector<int> dist(n, INF);
    vector<bool> inQueue(n, false);
    vector<int> cnt(n, 0);  // Number of times enqueued
    queue<int> q;

    dist[start] = 0;
    q.push(start);
    inQueue[start] = true;
    cnt[start] = 1;

    while (!q.empty()) {
        int u = q.front();
        q.pop();
        inQueue[u] = false;

        for (auto [v, w] : adj[u]) {
            if (dist[u] + w < dist[v]) {
                dist[v] = dist[u] + w;
                if (!inQueue[v]) {
                    q.push(v);
                    inQueue[v] = true;
                    cnt[v]++;

                    if (cnt[v] >= n) {
                        return {dist, true};  // Negative cycle
                    }
                }
            }
        }
    }

    return {dist, false};
}

// =============================================================================
// 6. K-th Shortest Path
// =============================================================================

vector<int> kthShortestPath(int n, const vector<vector<pair<int, int>>>& adj,
                            int start, int end, int k) {
    vector<int> kthDist;
    vector<int> count(n, 0);
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<>> pq;

    pq.push({0, start});

    while (!pq.empty() && count[end] < k) {
        auto [d, u] = pq.top();
        pq.pop();

        count[u]++;

        if (u == end) {
            kthDist.push_back(d);
        }

        if (count[u] <= k) {
            for (auto [v, w] : adj[u]) {
                pq.push({d + w, v});
            }
        }
    }

    return kthDist;
}

// =============================================================================
// Test
// =============================================================================

void printVector(const vector<int>& v) {
    cout << "[";
    for (size_t i = 0; i < v.size(); i++) {
        if (v[i] == INF) cout << "INF";
        else cout << v[i];
        if (i < v.size() - 1) cout << ", ";
    }
    cout << "]";
}

int main() {
    cout << "============================================================" << endl;
    cout << "Shortest Path Examples" << endl;
    cout << "============================================================" << endl;

    // Test graph
    int n = 5;
    vector<vector<pair<int, int>>> adj(n);
    adj[0] = {{1, 4}, {2, 1}};
    adj[1] = {{3, 1}};
    adj[2] = {{1, 2}, {3, 5}};
    adj[3] = {{4, 3}};

    // 1. Dijkstra
    cout << "\n[1] Dijkstra's Algorithm" << endl;
    auto dijkDist = dijkstra(n, adj, 0);
    cout << "    Distance from 0 to each vertex: ";
    printVector(dijkDist);
    cout << endl;

    // Path reconstruction
    auto [dist, parent] = dijkstraWithPath(n, adj, 0);
    auto path = reconstructPath(0, 4, parent);
    cout << "    0 -> 4 path: ";
    printVector(path);
    cout << " (distance: " << dist[4] << ")" << endl;

    // 2. Bellman-Ford
    cout << "\n[2] Bellman-Ford Algorithm" << endl;
    vector<Edge> edges = {
        {0, 1, 4}, {0, 2, 1}, {1, 3, 1}, {2, 1, 2}, {2, 3, 5}, {3, 4, 3}
    };
    auto [bfDist, hasNegCycle] = bellmanFord(n, edges, 0);
    cout << "    Distance from 0 to each vertex: ";
    printVector(bfDist);
    cout << endl;
    cout << "    Negative cycle: " << (hasNegCycle ? "exists" : "none") << endl;

    // 3. Floyd-Warshall
    cout << "\n[3] Floyd-Warshall Algorithm" << endl;
    vector<vector<int>> distMatrix(4, vector<int>(4, INF));
    for (int i = 0; i < 4; i++) distMatrix[i][i] = 0;
    distMatrix[0][1] = 3;
    distMatrix[0][2] = 8;
    distMatrix[1][2] = 2;
    distMatrix[2][3] = 1;
    distMatrix[0][3] = 7;

    floydWarshall(4, distMatrix);
    cout << "    All-pairs shortest distances:" << endl;
    for (int i = 0; i < 4; i++) {
        cout << "      " << i << ": ";
        printVector(distMatrix[i]);
        cout << endl;
    }

    // 4. 0-1 BFS
    cout << "\n[4] 0-1 BFS" << endl;
    vector<vector<pair<int, int>>> adj01(4);
    adj01[0] = {{1, 0}, {2, 1}};
    adj01[1] = {{2, 0}, {3, 1}};
    adj01[2] = {{3, 0}};
    auto dist01 = bfs01(4, adj01, 0);
    cout << "    Distance from 0 (0-1 weights): ";
    printVector(dist01);
    cout << endl;

    // 5. K-th Shortest Path
    cout << "\n[5] K-th Shortest Path" << endl;
    auto kthDists = kthShortestPath(n, adj, 0, 4, 3);
    cout << "    0 -> 4, 1st~3rd shortest distances: ";
    printVector(kthDists);
    cout << endl;

    // 6. Complexity Summary
    cout << "\n[6] Complexity Summary" << endl;
    cout << "    | Algorithm        | Time             | Neg. Weights |" << endl;
    cout << "    |------------------|------------------|--------------|" << endl;
    cout << "    | Dijkstra         | O((V+E) log V)   | No           |" << endl;
    cout << "    | Bellman-Ford     | O(VE)            | Yes (no neg.)|" << endl;
    cout << "    | Floyd-Warshall   | O(V^3)           | Yes (no neg.)|" << endl;
    cout << "    | 0-1 BFS          | O(V + E)         | 0,1 only     |" << endl;
    cout << "    | SPFA             | O(VE) avg O(E)   | Yes (no neg.)|" << endl;

    cout << "\n============================================================" << endl;

    return 0;
}
